plutarch-test/src/Plutarch/Test/Golden.hs

```diff
@@ -0,0 +1,184 @@
+{-# LANGUAGE ImpredicativeTypes #-}
+
+module Plutarch.Test.Golden (
+  pgoldenSpec,
+  (@|),
+  (@\),
+  (@->),
+
+  -- * Internal
+  TermExpectation,
+  goldenKeyString,
+  evaluateScriptAlways,
+  compileD,
+) where
+
+import Control.Monad (forM_, unless)
+import qualified Data.Aeson.Text as Aeson
+import Data.Kind (Type)
+import Data.List.NonEmpty (nonEmpty)
+import qualified Data.List.NonEmpty as NE
+import Data.Maybe (mapMaybe)
+import Data.Semigroup (sconcat)
+import Data.String (IsString)
+import Data.Text (Text)
+import qualified Data.Text as T
+import qualified Data.Text.Lazy as TL
+import GHC.Stack (HasCallStack)
+import System.FilePath ((</>))
+import Test.Syd (
+  Expectation,
+  Spec,
+  TestDefM,
+  describe,
+  getTestDescriptionPath,
+  it,
+  pureGoldenTextFile,
+ )
+
+import Plutarch
+import Plutarch.Benchmark (benchmarkScript')
+import Plutarch.Evaluate (evaluateScript)
+import Plutarch.Internal (Term (Term, asRawTerm))
+import Plutarch.Test.ListSyntax (ListSyntax, listSyntaxAdd, listSyntaxAddSubList, runListSyntax)
+import qualified Plutus.V1.Ledger.Scripts as Scripts
+
+data GoldenValue = GoldenValue
+  { goldenValueUplcPreEval :: Text
+  , goldenValueUplcPostEval :: Text
+  , goldenValueBench :: Text
+  , goldenValueExpectation :: Maybe Expectation
+  }
+
+{- Class of types that represent `GoldenValue`
+
+  This class exists for syntatic sugar provided by (@->) (via `TermExpectation`).
+-}
+class HasGoldenValue a where
+  mkGoldenValue :: a -> GoldenValue
+
+mkGoldenValue' :: ClosedTerm a -> Maybe Expectation -> GoldenValue
+mkGoldenValue' p mexp =
+  GoldenValue
+    (T.pack $ printScript $ compileD p)
+    (T.pack $ printScript $ evaluateScriptAlways $ compileD p)
+    (TL.toStrict $ Aeson.encodeToLazyText $ benchmarkScript' $ compileD p)
+    mexp
+
+instance HasGoldenValue (Term s a) where
```

Comment:
This is really bad IMO. This will fail at run-time when you try to use it on non-closed terms.
Can I ask why you need this?

Reply:
It exists mainly to provide the optional "And then check that this Plutarch program matches this expectation" syntax

```
    "por" @\ do
      "tf" @| pcon PTrue #|| pcon PFalse @-> passert
```

That `@-> passert`, basically. The type class treats the following (that comes after `@|`) polymorphically):

- `pcon PTrue #|| pcon PFalse`
- `pcon PTrue #|| pcon PFalse @-> \p -> passert p`

---

> This will fail at run-time when you try to use it on non-closed terms.

IIRC GHC prevented me from deriving for `ClosedTerm a`, forcing me to use `Term s a` ... let me check this once again.

Reply:
Yup, I can't do this:

<img width="503" alt="image" src="https://user-images.githubusercontent.com/3998/155029558-ad9de1c3-775d-4e7f-9f88-656e5d71d510.png">

Reply:
If you can make `@|` and such take closed terms and not export the instance, then it should be safe actually.

Reply:
Okay, should be good now. `ClosedTerm` everywhere except in this instance and `unsafeClosedTerm`.